File: packages/sdkwork-utils-java/src/main/java/com/sdkwork/utils/I18nUtils.java
package com.sdkwork.utils;

import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;

public final class I18nUtils {
    private I18nUtils() {
    }

    public static String formatNumberLocale(double value, String locale, int decimals) {
        boolean negative = value < 0;
        double absolute = Math.abs(NumberUtils.round(value, decimals));
        String decimalSeparator = "de-DE".equalsIgnoreCase(locale) ? "," : ".";
        String groupingSeparator = "de-DE".equalsIgnoreCase(locale) ? "." : ",";
        String formatted = String.format("%." + decimals + "f", absolute);
        String[] parts = formatted.split("\\.");
        String grouped = insertGrouping(parts[0], groupingSeparator);
        if (decimals > 0) {
            grouped = grouped + decimalSeparator + parts[1];
        }
        return negative ? "-" + grouped : grouped;
    }

    public static String formatDatetimeLocale(Instant value, String locale) {
        String normalized = locale == null ? "en-us" : locale.toLowerCase();
        DateTimeFormatter formatter;
        if (normalized.startsWith("de")) {
            formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy HH:mm").withZone(ZoneOffset.UTC);
        } else if (normalized.startsWith("zh")) {
            formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm").withZone(ZoneOffset.UTC);
        } else {
            formatter = DateTimeFormatter.ofPattern("MM/dd/yyyy HH:mm").withZone(ZoneOffset.UTC);
        }
        return formatter.format(value);
    }

    private static String insertGrouping(String integer, String groupingSeparator) {
        StringBuilder builder = new StringBuilder();
        int count = 0;
        for (int index = integer.length() - 1; index >= 0; index--) {
            if (count > 0 && count % 3 == 0) {
                builder.insert(0, groupingSeparator);
            }
            builder.insert(0, integer.charAt(index));
            count++;
        }
        return builder.toString();
    }

    public static Double parseNumberLocale(String input, String locale) {
        if (input == null) {
            return null;
        }
        String trimmed = input.trim();
        if (trimmed.isEmpty()) {
            return null;
        }
        String decimalSeparator = "de-DE".equalsIgnoreCase(locale) ? "," : ".";
        String groupingSeparator = "de-DE".equalsIgnoreCase(locale) ? "." : ",";
        String normalized = trimmed.replace(groupingSeparator, "").replace(decimalSeparator, ".");
        return NumberUtils.parseNumber(normalized);
    }
}

File: packages/sdkwork-utils-kotlin/src/main/kotlin/com/sdkwork/utils/I18nUtils.kt
package com.sdkwork.utils

import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

object I18nUtils {
    fun formatNumberLocale(value: Double, locale: String, decimals: Int = 0): String {
        val negative = value < 0
        val absolute = kotlin.math.abs(NumberUtils.round(value, decimals))
        val decimalSeparator = if (locale.equals("de-DE", ignoreCase = true)) "," else "."
        val groupingSeparator = if (locale.equals("de-DE", ignoreCase = true)) "." else ","
        val formatted = "%.${decimals}f".format(absolute)
        val parts = formatted.split(".")
        val grouped = insertGrouping(parts[0], groupingSeparator)
        val result = if (decimals > 0) grouped + decimalSeparator + parts[1] else grouped
        return if (negative) "-$result" else result
    }

    fun formatDatetimeLocale(value: Instant, locale: String): String {
        val normalized = locale.lowercase()
        val formatter = when {
            normalized.startsWith("de") -> DateTimeFormatter.ofPattern("dd.MM.yyyy HH:mm").withZone(ZoneOffset.UTC)
            normalized.startsWith("zh") -> DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm").withZone(ZoneOffset.UTC)
            else -> DateTimeFormatter.ofPattern("MM/dd/yyyy HH:mm").withZone(ZoneOffset.UTC)
        }
        return formatter.format(value)
    }

    private fun insertGrouping(integer: String, groupingSeparator: String): String {
        val builder = StringBuilder()
        var count = 0
        for (index in integer.indices.reversed()) {
            if (count > 0 && count % 3 == 0) builder.insert(0, groupingSeparator)
            builder.insert(0, integer[index])
            count++
        }
        return builder.toString()
    }

    fun parseNumberLocale(input: String, locale: String): Double? {
        val trimmed = input.trim()
        if (trimmed.isEmpty()) {
            return null
        }
        val decimalSeparator = if (locale.equals("de-DE", ignoreCase = true)) "," else "."
        val groupingSeparator = if (locale.equals("de-DE", ignoreCase = true)) "." else ","
        val normalized = trimmed.replace(groupingSeparator, "").replace(decimalSeparator, ".")
        return NumberUtils.parseNumber(normalized)
    }
}
